plugins/modules/ec2_instance.py

```diff
@@ -2066,343 +1897,156 @@ def enforce_count(existing_matches, desired_module_state):
-        module.deprecate(
-            "The tenancy parameter has been deprecated, please use placement.tenancy instead.",
-            date="2025-12-01",
-            collection_name="amazon.aws",
-        )
 
-    state = module.params.get("state")
+def run_ec2_instances(module: EC2InstanceModule, **instance_spec) -> Dict[str, Any]:
+    try:
+        result = run_instances(module.ec2, **instance_spec)
+    except AnsibleEC2Error as e:
+        if e.is_boto3_error_message("Invalid IAM Instance Profile ARN"):
+            # If the instance profile has just been created, it takes some time to be visible by ec2
+            # So we wait 10 second and retry the run_instances
+            time.sleep(10)
+            return run_instances(module.ec2, **instance_spec)
+        module.fail_json_aws(e, msg="Failed to run instances")
+    return result
 
-    retry_decorator = AWSRetry.jittered_backoff(
-        catch_extra_error_codes=[
-            "IncorrectState",
-            "InsuffienctInstanceCapacity",
-            "InvalidInstanceID.NotFound",
-        ]
-    )
-    client = module.client("ec2", retry_decorator=retry_decorator)
+
+def main() -> None:
+    module = EC2InstanceModule()
 
     filters = module.params.get("filters")
     if filters is None:
-        filters = build_filters()
+        filters = module.build_filters()
 
     try:
-        existing_matches = find_instances(filters=filters)
+        existing_matches = module.find_instances(filters=filters)
+        state = module.params.get("state")
 
         if state in ("terminated", "absent"):
-            if existing_matches:
-                result = ensure_instance_state(state, filters)
-            else:
-                result = dict(
-                    msg="No matching instances found",
-                    changed=False,
-                )
+            result = ensure_instance_state(module, existing_matches, filters)
         elif module.params.get("exact_count"):
-            result = enforce_count(existing_matches, desired_module_state=state)
+            result = enforce_count(module, existing_matches)
         elif existing_matches and not module.params.get("count"):
             for match in existing_matches:
-                warn_if_public_ip_assignment_changed(match)
-                warn_if_cpu_options_changed(match)
-            result = handle_existing(existing_matches, state, filters=filters)
+                warn_if_public_ip_assignment_changed(module, match)
+                warn_if_cpu_options_changed(module, match)
+            result = handle_existing(module, existing_matches, filters=filters)
         else:
-            result = ensure_present(existing_matches=existing_matches, desired_module_state=state)
+            result = ensure_present(module=module, existing_matches=existing_matches, desired_module_state=state)
     except Ec2InstanceAWSError as e:
         if e.exception:
             module.fail_json_aws(e.exception, msg=e.message)
```

Comment:
This (and the `fail_json(msg=e.message)` line below) can now be simplified to

```suggestion
    except AnsibleAWSError as e:
        module.fail_json_aws_error(e)
```